### PR TITLE
Add support for managing gtm f5 resources

### PR DESCRIPTION
**Pull request by mabuaisha**

### CODE REVIEW

File: plugin.yaml [hunk 1]
```diff
@@ -46,3 +92,90 @@ node_types:
...
+           pool_member_config:
+             default: { get_property: [SELF, pool_member_config]}
+           wipe_config:
+             default: { get_property: [SELF, wipe_config]}
+        delete:
+          implementation: bigip.bigip_plugin.bigip_gtm.delete_dns_loadbalancer
+
+relationships:
+
+  cloudify.bigip.gtm.relationships.managed_by_gtm:
```
Comment (EarthmanT) [inline plugin.yaml]: What is the point of having a special relationship if there is no operation mapping?

File: plugin.yaml [hunk 2]
```diff
@@ -46,3 +92,90 @@ node_types:
...
+        description: >
+          IP of the BIG IP GTM instance
+      user:
+        type: string
+        description: >
+          User used to login to BIG IP GTM instance
+      password:
+        type: string
+        description: >
+          Password used to login to BIG IP GTM instance
```
Comment (EarthmanT) [inline plugin.yaml]: Why not have this be a property of the `cloudify.nodes.bigip.gtm.DNSLoadBalancer` node type?

File: bigip_plugin/bigip_gtm.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    ctx.logger.info("Adding Virtual Server ={0} To GTM Server={1}, "
+                    "Partition={2}".format(vs_name, ltm_server, partition))
+
+    # Update the name of the virtual server
+    resource['name'] = '/{0}/{1}'.format(partition, vs_name)
+    response = igtm.add_vs(partition, ltm_server, resource)
+
+    # Log created response
+    ctx.logger.info("Adding Virtual Server ={0}"
+                    " Created Successfully {1} ".format(vs_name, response))
```
Comment (EarthmanT) [inline bigip_plugin/bigip_gtm.py]: Do we need the space after `{1} `?

File: bigip_plugin/utils.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+def with_gtm_client(function):
+    def wrapper(**kwargs):
+        # This will fetch the properties of the GTM instance from the
+        # relationship of the node
+        try:
+            properties = _retrieve_properties(ctx.instance)
+            gmt_prop = {k: properties[k] for k in GTM_KEYS}
+            kwargs['igtm'] = gtm.GTM(**gmt_prop)
+            function(**kwargs)
+        except Exception as error:
```
Comment (EarthmanT) [inline bigip_plugin/utils.py]: Do we want all exceptions to be treated equally?

File: bigip_plugin/utils.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+            error_traceback = generate_traceback_exception()
+            error_message = 'Error traceback {0} with message' \
+                            ' {1}'.format(error_traceback['traceback'],
+                                          error_traceback['message'])
+            ctx.logger.error(error_message)
+            raise NonRecoverableError('{0}'.format(str(error)),
+                                      causes=[error_traceback])
```
Comment (EarthmanT) [inline bigip_plugin/utils.py]: Are there any "recoverable" or "retry" errors?

**Review by EarthmanT — APPROVED** [review]
Looks good. Left some questions.

File: plugin.yaml (hunk 1 above)
Comment (mabuaisha) [reply plugin.yaml]: Well, this kind of instance will be act as config node in order to use the credential in order to connect to the GTM servers, and In case we need to add any additional resources we do not need then to duplicate the credential every-time we need to add new type.

You can think of this like K8S master node which is user to hold the config for the all data required to any created node to connect to the cluster

File: plugin.yaml (hunk 2 above)
Comment (mabuaisha) [reply plugin.yaml]: As I mentioned before, we may need to add additional node types and each node type then will required to add these information each time a node is created

File: bigip_plugin/bigip_gtm.py (hunk 3 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_gtm.py]: Done

File: bigip_plugin/utils.py (hunk 4 above)
Comment (mabuaisha) [reply bigip_plugin/utils.py]: I thought about this, and since we have a single node that do all the operations, adding another exceptions handling to cause a retry or recoverable could lead to create another resources already created  which raise another exception which will cause another complexity of the system to handle already created resources.

Using multiple nodes save us from this issue, but the asked to combine all the operations in one node. and based on that I added the built it

File: bigip_plugin/utils.py (hunk 5 above)
Comment (mabuaisha) [reply bigip_plugin/utils.py]: See above response

File: plugin.yaml (hunk 1 above)
Comment (EarthmanT) [reply plugin.yaml]: We are planning to get rid of the way we do it in the Kubernetes plugin, because it is not as good as a property

Comment (mabuaisha) [reply plugin.yaml]: Fixed and removed

File: README.md [hunk 6]
```diff
@@ -26,3 +26,42 @@ type_names: [cloudify.bigip.BigIP]
 ```
 
 This will invoke the `cluster.operations.sync` operation on all node templates in the blueprint that belong to the type `cloudify.big.BigIP`. You can also specify `node_ids` instead of `type_names` in order to address a particular node template; see the documentation for the `execute_operation` built-in workflow.
+
+
+### `cloudify.nodes.bigip.gtm.DNSLoadBalancer`
+
+Represents a DNS Load Balancer that responsible for creating the following resources:
```
Comment (isaac-s) [inline README.md]: "that responsible" -> "that is responsible"

File: bigip_plugin/bigip_cert.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+@operation
+def install_cert(key, cert, ctx):
+
+    ip = ctx.node.properties['ip']
+    user = ctx.node.properties['user']
+    password = ctx.node.properties['password']
+    name = ctx.node.properties['name']
+
+    # save certs to file for http access
+    base_path = "/opt/manager/resources/cloudify_agent"
```
Comment (isaac-s) [inline bigip_plugin/bigip_cert.py:38]: We should never rely on such file paths.
This makes the plugin sensitive to the physical filesystem layout of the manager.
Why can't this be written to `/tmp` (or whatever `gettempdir()` returns)?

File: bigip_plugin/bigip_cert.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    )
+    _write_temp_file(os.path.join(base_path, key_filename), key)
+
+    ctx.logger.info(
+        'Generating cert file {name} on manager'.format(name=cert_filename)
+    )
+    _write_temp_file(os.path.join(base_path, cert_filename), cert)
+
+    mgr_ip = _get_active_manager_ip(ctx)
+    base_url = '{ip}:53229/cloudify_agent/{filename}'
```
Comment (isaac-s) [inline bigip_plugin/bigip_cert.py:53]: We need a more robust way of obtaining this.
You must never hard-code such ports or paths.
Please ask R&D for help — specifically Ido or @tehasdf.

File: bigip_plugin/bigip_cert.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    key_url = base_url.format(
+        ip=mgr_ip,
+        filename=key_filename
+    )
+    cert_url = base_url.format(
+        ip=mgr_ip,
+        filename=cert_filename
+    )
+
+    ctx.logger.info("Installing certificate {0} on BIG-IP={1}"
```
Comment (isaac-s) [inline bigip_plugin/bigip_cert.py:63]: "Installing certificate {0} on BIG-IP located at {1}"

File: bigip_plugin/bigip_cert.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+                    .format(name, ip))
+    try:
+        result = bigip.install_cert(ip,
```
Comment (isaac-s) [inline bigip_plugin/bigip_cert.py:66]: Can we avoid using so many lines? code is less readable this way. No need to break a function call over so many lines.

File: bigip_plugin/bigip_cert.py [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+                                    name,
+                                    key_url,
+                                    cert_url)
+
+    except Exception as e:
+        _, exc_value, exc_traceback = sys.exc_info()
+        raise NonRecoverableError(
+            "Failed to install certificate on {load_balancer}. "
+            "{exception_type} {message}".format(
+                load_balancer=ip,
```
Comment (isaac-s) [inline bigip_plugin/bigip_cert.py:78]: We don't need the exception's type and message as a part of the `NonRecoverableError`'s message. They will be printed already along with the `causes` thing. Let's avoid double-logging of data.

File: bigip_plugin/bigip_cert.py [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+                exception_type=type(e),
+                message=e.message,
+            ),
+            causes=[exception_to_error_cause(exc_value, exc_traceback)]
+        )
+
+    finally:
+        # remove temp files
+        ctx.logger.info('Cleaning up cert and key files.')
+        os.remove(os.path.join(base_path, key_filename))
```
Comment (isaac-s) [inline bigip_plugin/bigip_cert.py:88]: `os.remove` may actually raise an exception in certain cases.

File: bigip_plugin/bigip_cert.py [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+
+from cloudify.decorators import operation
+from cloudify.exceptions import NonRecoverableError
+from cloudify.utils import exception_to_error_cause
+from cloudify import cluster, constants
+
+from bigip_sdk import bigip
+
+
+@operation
```
Comment (isaac-s) [inline bigip_plugin/bigip_cert.py:29]: I'm pretty sure we don't need `@operation` anymore.
@EarthmanT @0lvin can you confirm?

File: bigip_plugin/bigip_cert.py [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+    ip = ctx.node.properties['ip']
+    user = ctx.node.properties['user']
+    password = ctx.node.properties['password']
+
+    key_name = ctx.instance.runtime_properties['key_name']
+    cert_name = ctx.instance.runtime_properties['cert_name']
+
+    ctx.logger.info(
+        'Deleting key {key_name} and cert {cert_name} from '
+        '{load_balancer}.'.format(
```
Comment (isaac-s) [inline bigip_plugin/bigip_cert.py:106]: It is a convention (AFAIK) to not end log lines with a dot.

File: bigip_plugin/bigip_cert.py [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+
+import sys
+import os
+import uuid
+from shutil import copyfile
+
+from cloudify.decorators import operation
+from cloudify.exceptions import NonRecoverableError
+from cloudify.utils import exception_to_error_cause
+from cloudify import cluster, constants
```
Comment (isaac-s) [inline bigip_plugin/bigip_cert.py:24]: Please confirm with R&D that importing `cloudify.cluster`, and using its functions (as done below), is safe and supported for plugins.

File: bigip_plugin/bigip_cert.py [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+            load_balancer=ip,
+        )
+    )
+
+
+def _get_active_manager_ip(ctx):
+    if cluster.is_cluster_configured():
+        active_node_ip = cluster.get_cluster_active()
+        if active_node_ip:
+            ctx.logger.debug("active node ip")
```
Comment (isaac-s) [inline bigip_plugin/bigip_cert.py:133]: A better print would be:

`ctx.logger.debug("Active node IP response: {0}".format(active_node_ip))`

File: bigip_plugin/bigip_cert.py [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+            return str(active_node_ip)[:-16]
```
Comment (isaac-s) [inline bigip_plugin/bigip_cert.py:134]: Why are we using string manipulation here to extract the data?
`16` also looks like some sort of a magic number. I'm hypothesizing that this has something to do with `len('aaa.bbb.ccc.ddd')` being 15 and I'm not sure where the extra character goes.
Regardless, there must be a better way to obtain the IP.

File: bigip_plugin/bigip_cert.py [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+    ctx.logger.debug("from OS active node ip")
+    return str(os.environ[constants.MANAGER_FILE_SERVER_URL_KEY])[:-16]
+
+
+def _write_temp_file(filename, source):
+    if os.path.isfile(source):
+        copyfile(source, filename)
+    else:
+        with open(filename, "w") as key_handle:
```
Comment (isaac-s) [inline bigip_plugin/bigip_cert.py:143]: Shouldn't the open mode be `wb`?

File: bigip_plugin/bigip_gtm.py [hunk 19]
```diff
@@ -0,0 +1,236 @@
...
+from cloudify.exceptions import NonRecoverableError
+
+# local Imports
+from . import utils
+
+
+def _create_virtual_server(ctx, igtm, partition, gtm_server, resource_config):
+    vs_name = resource_config.get('name')
+
+    ctx.logger.info("Adding Virtual Server ={0} To GTM Server={1}, "
```
Comment (isaac-s) [inline bigip_plugin/bigip_gtm.py]: Avoid the `=` sign.
Please correct capitalization.

`Adding virtual server {0} to GTP server located at {1}`

File: bigip_plugin/bigip_gtm.py [hunk 20]
```diff
@@ -0,0 +1,236 @@
...
+                    "Partition={2}".format(vs_name, gtm_server, partition))
+
+    # Update the name of the virtual server
+    resource_config['name'] = '/{0}/{1}'.format(partition, vs_name)
+    response = igtm.add_vs(partition, gtm_server, resource_config)
+
+    # Log created response
+    ctx.logger.info("Adding Virtual Server ={0}"
```
Comment (isaac-s) [inline bigip_plugin/bigip_gtm.py]: See above re: formatting and grammar.

File: bigip_plugin/bigip_gtm.py [hunk 21]
```diff
@@ -0,0 +1,236 @@
...
+        created_servers.append(response)
+
+    if created_servers:
+        # Set runtime properties for created virtual servers instances
+        ctx.instance.runtime_properties['virtual_servers'] = created_servers
+
+
+def _delete_virtual_servers(ctx, igtm):
+    virtual_servers = ctx.instance.runtime_properties.get('virtual_servers')
+    # Check if there is an object called ``virtual_servers`` so that the
```
Comment (isaac-s) [inline bigip_plugin/bigip_gtm.py]: Replace "``" with regular quotes.

File: bigip_plugin/bigip_gtm.py [hunk 22]
```diff
@@ -0,0 +1,236 @@
...
+    # resources can be removed when running the un-install workflow
+    if not virtual_servers:
+        raise NonRecoverableError('No Virtual Servers Found To Delete')
```
Comment (isaac-s) [inline bigip_plugin/bigip_gtm.py]: Capitalization.

File: bigip_plugin/bigip_gtm.py [hunk 23]
```diff
@@ -0,0 +1,236 @@
...
+                                    gtm_server, pool_member)
+
+        added_members.append(response)
+
+    if added_members:
+        # Set runtime properties for added virtual pool members
+        ctx.instance.runtime_properties['pool_members'] = added_members
+
+
+def _create_wipe(ctx, igtm, record_type, resource_config):
```
Comment (isaac-s) [inline bigip_plugin/bigip_gtm.py]: "wipe"?
Shouldn't this be "wide"?

File: bigip_plugin/bigip_gtm.py [hunk 24]
```diff
@@ -0,0 +1,236 @@
...
+                            igtm,
+                            partition,
+                            gtm_server,
+                            record_type,
+                            virtual_server_config,
+                            pool_config,
+                            pool_member_config,
+                            wipe_config):
+
+    # Create multiple virtual servers
```
Comment (isaac-s) [inline bigip_plugin/bigip_gtm.py]: Possibility of resource leak in case the operation encounters an error and being retried.
Use `ctx.operation.retry_number` to guard against this, and use interim runtime properties to make sure you don't re-do things that you shouldn't re-do.

File: bigip_plugin/bigip_sync.py [hunk 25]
```diff
@@ -41,8 +41,14 @@ def sync(sync_group,
                         .format(ip, sync_group))
 
         bigip.do_sync(ip, sync_group, user, password, retry_timer)
-    except Exception:
+    except Exception as e:
         _, exc_value, exc_traceback = sys.exc_info()
         raise NonRecoverableError(
-            "Failed syncing BIG IP with ip '{0}'".format(ip),
-            causes=[exception_to_error_cause(exc_value, exc_traceback)])
+            "Failed to install certificate on {load_balancer}. "
```
Comment (isaac-s) [inline bigip_plugin/bigip_sync.py:47]: Double logging of error message.

**Review by isaac-s — CHANGES_REQUESTED** [review]

File: bigip_plugin/bigip_cert.py (hunk 15 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_cert.py:24]: @isaac-s I forgot to mentioned that the whole new module `bigip_cert.py` was not part of the `cloudify-bigip-plugin` on cloudify repo, so I just added it here, to make it up to date with FICO code, so basically I just copy/paste the module and this is has nothing to do with GTM plugin 

This `https://gitserver.fairisaac.com:8443/projects/PLUGINS/repos/bigip-plugin/browse` version of bigip-plugins does not match the one we already have on `Cloudify-PS` so I had to copy/update the missing file which are added before.

Basically, This is not the code I implemented  and I do no have enough information about the cert module for bigip.

File: bigip_plugin/bigip_cert.py (hunk 13 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_cert.py:29]: @isaac-s I'm not sure how to test `bigip_cert.py` module and I I'm not sure who implement this part

What do you think about  skip the changes you requested for this module to avoid breaking anything because honestly I do not have details about cert in bigip

File: bigip_plugin/bigip_cert.py (hunk 7 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_cert.py:38]: @isaac-s As I told you before I do not have enough info about this module, just want to have the code for bigip-plugins compatible with the FICO version

File: bigip_plugin/bigip_cert.py (hunk 8 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_cert.py:53]: @isaac-s check above

File: bigip_plugin/bigip_sync.py (hunk 25 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_sync.py:47]: @isaac-s This change because of the FICO `https://gitserver.fairisaac.com:8443/projects/PLUGINS/repos/bigip-plugin/browse`

File: bigip_plugin/bigip_gtm.py (hunk 23 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_gtm.py]: This is called `wipe` which is a F5 resource inside the GTM

File: bigip_plugin/bigip_gtm.py (hunk 21 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_gtm.py]: Done

File: bigip_plugin/bigip_cert.py (hunk 9 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_cert.py:63]: @isaac-s check above

File: bigip_plugin/bigip_cert.py (hunk 10 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_cert.py:66]: @isaac-s check above

File: bigip_plugin/bigip_cert.py (hunk 11 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_cert.py:78]: @isaac-s check above

File: bigip_plugin/bigip_cert.py (hunk 12 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_cert.py:88]: @isaac-s check above

File: bigip_plugin/bigip_cert.py (hunk 14 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_cert.py:106]: @isaac-s check above

File: bigip_plugin/bigip_cert.py (hunk 16 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_cert.py:133]: @isaac-s check above

File: bigip_plugin/bigip_cert.py (hunk 17 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_cert.py:134]: @isaac-s check above

File: bigip_plugin/bigip_cert.py (hunk 18 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_cert.py:143]: @isaac-s check above

File: bigip_plugin/bigip_gtm.py (hunk 19 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_gtm.py]: Done

File: bigip_plugin/bigip_gtm.py (hunk 20 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_gtm.py]: Done

File: bigip_plugin/bigip_gtm.py (hunk 22 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_gtm.py]: Done

File: bigip_plugin/bigip_gtm.py (hunk 24 above)
Comment (mabuaisha) [reply bigip_plugin/bigip_gtm.py]: @isaac-s Good catch, I will update this

File: README.md (hunk 6 above)
Comment (mabuaisha) [reply README.md]: Done
